components/proxy/src/test/java/com/hotels/styx/proxy/backends/file/FileChangeMonitorTest.java

```diff
@@ -0,0 +1,149 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hotels.styx.proxy.backends.file;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import static com.google.common.io.Files.createTempDir;
+import static java.lang.String.format;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.nio.file.Files.copy;
+import static java.nio.file.Files.delete;
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.timeout;
+import static org.mockito.Mockito.verify;
+
+
+public class FileChangeMonitorTest {
+    private static final Logger LOGGER = LoggerFactory.getLogger(FileChangeMonitorTest.class);
+
+    private File tempDir;
+    private Path monitoredFile;
+    private FileMonitor.Listener listener;
+    private FileChangeMonitor monitor;
+
+    @BeforeMethod
+    public void setUp() throws Exception {
+        tempDir = createTempDir();
+        monitoredFile = Paths.get(tempDir.toString(), "origins.yml");
+        write(monitoredFile, "content-v0");
+        listener = mock(FileChangeMonitor.Listener.class);
+        monitor = new FileChangeMonitor(monitoredFile.toString(), 250, MILLISECONDS);
+    }
+
+    @AfterMethod
+    public void tearDown() throws Exception {
+        try {
+            delete(monitoredFile);
+        } catch (java.nio.file.NoSuchFileException cause) {
+            // Pass ...
+        }
+
+        delete(tempDir.toPath());
+    }
+
+    @Test(expectedExceptions = IllegalArgumentException.class)
+    public void throwExceptionIfFileDoesNotExist() {
+        new FileChangeMonitor("/nonexistant/file");
+    }
+
+    @Test(expectedExceptions = IllegalStateException.class, expectedExceptionsMessageRegExp = "File monitor for '.*' is already started")
+    public void canBeStartedOnlyOnce() {
+        FileChangeMonitor.Listener listener = mock(FileChangeMonitor.Listener.class);
+        FileChangeMonitor monitor = new FileChangeMonitor(monitoredFile.toString());
+
+        monitor.start(listener);
+        monitor.start(listener);
+    }
+
+    @Test
+    public void initialPollNotifiesListeners() {
+        monitor.start(listener);
+        verify(listener, timeout(3000).times(1)).fileChanged();
+    }
+
+
+    @Test
+    public void notifiesListenersOnFileChange() throws Exception {
+        monitor.start(listener);
+        verify(listener, timeout(3000).times(1)).fileChanged();
+
+        Thread.sleep(1000);
```

Comment:
We are adding a lot of sleeps in these tests, which all add up to make the total build time longer. Is there no way to avoid it?

Reply:
I have reduced sleeping. I couldn't think of any way of completely avoiding sleeps without more extensive changes. At the moment we are adding about 2.25 seconds to the overall build time.

Reply:
Yeah, I have a feeling in this particular case we may not be able to easily avoid it, but the reason we try to is that when we have lots of tests adding just a couple of seconds (added gradually over months/years), those seconds add up to minutes.
I think we'll just have to accept it for this class though.

Reply:
So, would it be acceptable in this case to simply change the modification time  manually (using Files.setLasModifiedTime() or File.setLastModified() )  and thus avoid having to sleep at all?